src/Stack/SDist.hs

```diff
@@ -93,71 +95,98 @@ instance Show CheckException where
+            | tweakCabal = do
                 (_ident, lbs) <- getCabalLbs pvpBounds Nothing cabalfp sourceMap
-                currTime <- liftIO getPOSIXTime -- Seconds from UNIX epoch
-                tp <- liftIO $ tarPath False fp
-                return $ (Tar.fileEntry tp lbs) { Tar.entryTime = floor currTime }
-            | otherwise = packWith packFileEntry False fp
+                curTime <- curModTime
+                let fi' =
+                        fi
+                            { CTar.fileModTime = curTime
+                            , CTar.fileSize = fromIntegral (L.length lbs)
+                            }
+                pure (fi', C.sourceLazy lbs)
+            | otherwise = pure (fi, C.sourceFile fp)
         isCabalFp fp = toFilePath pkgDir FP.</> fp == toFilePath cabalfp
-        tarName = pkgId FP.<.> "tar.gz"
         pkgId = packageIdentifierString (packageIdentifier (lpPackage lp))
-    dirEntries <- mapM packDir (dirsFromFiles files)
-    fileEntries <- mapM packFile files
+        yieldPackageFiles fs conf =
+            C.fuseUpstream
+                (C.yieldMany fs .| CTar.tarFiles conf)
+                (C.mapC (mapLeft resetFileInfo))
+        yieldCabalFile fp = do
+            fi <- resetFileInfo <$> CTar.getFileInfo fp
+            (fiAdjusted, cabalFileSource) <- lift $ lift $ getCabalFile fp fi
+            C.yield $
+                Left $
+                fiAdjusted
+                    { CTar.filePath =
+                          CTar.encodeFilePath (pkgId <> "/" <> FP.normalise fp)
+                    }
+            cabalFileSource .| C.mapC Right
+        sourcePackage =
+            case break isCabalFp files of
+                (_, []) -> lift $ lift $ logError "Could not find a .cabal file"
+                (filesBeforeCabal, cabalFile:filesAfterCabal) -> do
+                    pkgBaseDir <- CTar.createFTDirectory pkgId
+                    C.yield $ Left pkgBaseDir
+                    let tarFilesConfig =
+                            CTar.defaultTarFilesConfig
+                                { CTar.tarFilesBaseDirectory = Just pkgBaseDir
+                                , CTar.tarFilesDepth = Just 1
+                                , CTar.tarFilesDirectories =
+                                      Set.singleton
+                                          (CTar.getFileInfoPath pkgBaseDir)
+                                }
+                    createdDirs <-
+                        yieldPackageFiles filesBeforeCabal tarFilesConfig
+                    yieldCabalFile cabalFile
+                    void $
+                        yieldPackageFiles
+                            filesAfterCabal
+                            tarFilesConfig
+                                {CTar.tarFilesDirectories = createdDirs}
+        tarByteStream =
+            sourcePackage .| void CTar.tar .|
+            compress (-1) (WindowBits 31) -- switch to zlib's default compression.
```

Comment:
Switch to default gzip, this was fixed already: https://github.com/snoyberg/conduit/pull/406